src/Server/HTTPProxyProtocolHandler.cpp

```diff
@@ -18,36 +23,32 @@ void HTTPProxyProtocolHandler::handle(const HTTPServerRequest & request)
 {
     address_chain.clear();
 
+    std::vector<Poco::Net::IPAddress> addresses;
     std::vector<std::string> addresses_str;
     boost::split(addresses_str, request.get("X-Forwarded-For", ""), boost::is_any_of(","));
 
-    for (std::size_t i = 0; i < addresses_str.size(); ++i)
+    for (auto & address_str : addresses_str)
     {
-        auto & address_str = addresses_str[i];
         boost::trim(address_str);
+        addresses.emplace_back(address_str);
+    }
 
-        const auto client_address = (i == 0);
-        if (client_address)
+    std::size_t trim_front_n = 0;
+    if (config.proxy_chain_limit > 0 && config.proxy_chain_limit < addresses.size())
+        trim_front_n = addresses.size() - config.proxy_chain_limit;
+
+    for (std::size_t i = trim_front_n; i < addresses.size(); ++i)
+    {
+        // The first address is the client address, we don't need to verify it against trusted_networks.
+        if (i == trim_front_n || Util::addrInNet(addresses[i], config.trusted_networks))
         {
-            if (address_str.empty())
-            {
-                address_chain.clear();
-                return;
-            }
+            address_chain.push_back(std::move(addresses[i]));
         }
         else
         {
-            if (address_str.empty())
-                continue;
-
-            if (!Util::addrInNet(boost::asio::ip::make_address(address_str), config.trusted_networks))
-            {
-                address_chain.clear();
-                return;
-            }
+            address_chain.clear();
+            throw Exception("Untrusted proxy in the relevant chain of forwarded addresses", ErrorCodes::UNTRUSTED_PROXY);
```

Comment:
I thought you will just 'give deeper' while the address is trusted. With max allowed number of hops.

```
X-Forwarder-For: 127.0.0.1,3.3.3.3,10.0.0.1,10.0.0.2

trusted proxy: 10.0.0.0/24
proxy_chain_limit: 0 

10.0.0.2 - is trusted, go to the next 
10.0.0.1 - is trusted, go to the next 
3.3.3.3 - is not trusted, stop here, 3.3.3.3 is our address. 

trusted proxy: 10.0.0.0/24
proxy_chain_limit: 1

10.0.0.2 - is trusted, go to the next 
10.0.0.1 - proxy_chain_limit reached, stop here, 10.0.0.1 is our address.

trusted proxy: 10.0.0.0/24, 3.3.3.0/24
proxy_chain_limit: 10

10.0.0.2 - is trusted, go to the next 
10.0.0.1 - is trusted, go to the next 
3.3.3.3 - is trusted, go to the next 
127.0.0.1 - is the last address,  stop here, 127.0.0.1 is our address.
```

Reply:
Yes, that was the intended behavior of this code. Do you see any problems in it?

Reply:
trusted proxy: 10.0.0.0/24
proxy_chain_limit: 0 (all but the first address from left MUST be trusted)

10.0.0.2 - is trusted, go to the next
10.0.0.1 - is trusted, go to the next
3.3.3.3 - is not trusted, stop here, reject the request

other examples are correct.

Reply:
Why should we reject something on that level? If the proxy is not trusted - we don't care what is behind that.

Reply:
Let's review the use cases as we understand them...

(`ProxyN` is our immediate peer which is verified a priori.)
For `X-Forwarded-For: Client, Proxy1, Proxy2, ..., ProxyN-1`, server admin may want to:

1. get the true client address or nothing: all proxies in the chain must be verified, the last address from the right is the client address
2. get the first unverified (or the last from the right) address from the chain as a (possibly imprecise) client address

What is the configuration/limitation we want to support for case no. 2?

Reply:
> Mikhail Filimonov, [20.08.21 12:34]
> Про сетки. 
> 
> Исходное состояние: 
> 1. есть КХ сервер, и есть куча HTTP клиентов. Часть клиентов присылают всякие XFF (например используют локальный прокси для дебага / кэширования / отрезания рекламы, или "я у мамы хакер" и т.п.). Никаких конфигураций прокси у нас нет. КХ просто игнорирует все XFF и использует тот адрес который к нему подключился. 
> 
> 2. Теперь админ огранизации / КХ решает - что другой отдел той же огранизации надо подключить к КХ, но сетки разные и хочется дополнительный контроль. Добавляет в сетку прокси сервер, и добавляет его адрес как доверенный. Ставит proxy_chain_limit=1. 
> 
> При твоей реализации все те кто ранее использовал XFF (п.1) вдруг все перестает работать (Untrusted proxy). Очень странно и нелогично. Что теперь - надо всем HTTP клиентам которые присылают всякие XFF надо просить чтоб перестали присылать или добавлять их в доверенные?
> 
> Mikhail Filimonov, [20.08.21 12:49]
> Почему рекурсивно (на глубину больше одного прокси) небезопасно:
> 
> типичная ситуация: 
> есть внутренняя сетка. Пусть будет 10.1.1.0/24
> есть КХ внутри неё (адрес 10.1.1.100)
> у КХ есть клиент "босс" (адрес 10.1.1.1).
> доступ к этой сетке возможен только через прокси (адрес снаружи 200.200.200.1, адрес внутри 10.1.1.200) 
> КХ доверяет адресу 10.1.1.200 
> 
> Пока всё прекрасно. КХ теперь может обслуживать внешних клиентов, и видеть их внешние IP.
> 
> Теперь мы решаем расширить наше доверие. 
> 
> Пусть у нас есть ещё одна организация / или другой отдел в другом городе и т.д. И мы решаем что надо им предоставить доступ в наш КХ. 
> У них тоже внутренняя сетка 10.0.0.0/8 
> Они ходят наружу через прокси (наружний адрес 100.100.100.1, адрес внутри 10.3.3.300 )
> Мы добавляем их адрес (100.100.100.1) в список доверенных прокси. 
> 
> Тут начинаются проблемы:
> Внутри их сетки тоже есть адрес 10.1.1.1 который неожиданно получает права / квоты нашего босса. 
> Но ещё у них сетка очень широкая, и можно исплользовать любые статические адреса
> Хакер Петя знает конфигурацию нашего КХ и меняет адрес своей машины внутри их 10.1.1.200 (этот адрес в списке доверенных КХ).
> И теперь он со своей машины может отправлять любые XFF которые КХ будет считать валидными (так как суффикс содержит только адреса валидных прокси: 10.1.1.200, 100.100.100.1, 10.1.1.200

Reply:
This doesn't answer my question. I thought we discussed this and came to a conclusion about how to implement this.